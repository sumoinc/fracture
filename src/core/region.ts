import { FractureComponent } from ".";
import { Account } from "./account";

export interface RegionOptions {
  regionId: string;
}

export class Region extends FractureComponent {
  // member components
  // parent
  public readonly account: Account;
  // all other options
  public readonly options: RegionOptions;

  constructor(account: Account, options: RegionOptions) {
    super(account.fracture);

    /***************************************************************************
     *
     * INIT OPERATION
     *
     **************************************************************************/

    // member components

    // parents + inverse
    this.account = account;
    this.account.regions.push(this);

    // all other options
    this.options = options;

    this.project.logger.info(`INIT Region: "${this.regionId}"`);
  }

  public get regionId() {
    return this.options.regionId;
  }
}
