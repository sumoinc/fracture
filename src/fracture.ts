import { Project } from 'projen';
import { AppSync } from './appsync/appsync';
import { FractureComponent } from './component';
import { Entity, EntityOptions } from './model';

export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   */
  gendir?: string;
  /**
   * Enable AppSync
   */
  appsync?: boolean;
}

export class Fracture extends FractureComponent {
  public readonly gendir: string;
  public readonly appsync: boolean;

  constructor(
    project: Project,
    namespace?: string,
    options: FractureOptions = {},
  ) {
    super(project, namespace ?? 'fracture' );

    this.gendir = options.gendir ?? project.outdir;
    this.appsync = options.appsync ?? true;

    if (this.appsync) {
      new AppSync(project, this.namespace, { gendir: this.gendir });
    }
  }

  /**
   * Get all entities in this project's namespace.
   */
  public get entities(): Entity[] {
    const isEntity = (c: FractureComponent): c is Entity => c instanceof Entity && c.namespace === this.namespace;
    return (this.project.components as FractureComponent[]).filter(isEntity);
  }

  public addEntity(options: EntityOptions) {
    return new Entity(this.project, this.namespace, options);
  }


}
