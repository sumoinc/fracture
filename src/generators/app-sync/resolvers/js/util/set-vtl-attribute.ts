import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
