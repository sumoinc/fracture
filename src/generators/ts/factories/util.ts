import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
