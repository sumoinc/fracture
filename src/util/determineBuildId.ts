import { execSync } from "child_process";

/**
 * Returns the current git branch name, without folder.
 *
 * ie: feature/1234 returns 1234
 *
 * @returns {string}
 */
export const determineBuildId = () => {
  const branchName = execSync("git rev-parse --abbrev-ref HEAD")
    .toString("utf8")
    .replace(/[\n\r\s]+$/, "");
  const branchPath = branchName.split("/");
  return branchPath[branchPath.length - 1];
};
