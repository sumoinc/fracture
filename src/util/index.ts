export * from "./determineBuildId";
export * from "./test-util";
