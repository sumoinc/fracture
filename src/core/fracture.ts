import { Component, LoggerOptions, Project, LogLevel } from "projen";
import { deepMerge } from "projen/lib/util";
import { AuditStrategy } from "./audit-strategy";
import { NamingStrategy, NAMING_STRATEGY_TYPE } from "./naming-strategy";
import { OPERATION_SUB_TYPE } from "./operation";
import { Organization, OrganizationOptions } from "./organization";
import {
  ResourceAttributeGenerator,
  ResourceAttributeType,
} from "./resource-attribute";
import { Service, ServiceOptions } from "./service";
import { STRUCTURE_TYPE } from "./structure";
import { TypescriptService } from "../generators/ts/typescript-service";

/**
 *
 * Top level options for Fracture. These options also apply to services and
 * other sub components as they are created, unless they are overridden.
 *
 */
export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   * @default project.outdir + "/" + namespace
   */
  outdir?: string;
  /**
   * Logging options
   * @default LogLevel.INFO
   */
  logging?: LoggerOptions;
  /**
   * Versioned.
   * @default true
   */
  isVersioned?: boolean;
  /**
   * The naming strategy to use for generated code.
   */
  namingStrategy?: NamingStrategy;
  /**
   * The audit strategy to use for generated code.
   */
  auditStrategy?: AuditStrategy;
}

/**
 * The root of the entire application.
 */
export class Fracture extends Component {
  // member components
  public readonly services: Service[];
  public readonly organizations: Organization[];
  // project and namespace
  public readonly project: Project;
  public readonly namespace: string;
  // all other options
  public readonly options: Required<FractureOptions>;

  constructor(
    project: Project,
    namespace: string = "fracture",
    options: FractureOptions = {}
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     * These are the options that will be used through all code generation
     * unless explicitly overridden.
     *
     **************************************************************************/

    const defaultOptions: Required<FractureOptions> = {
      outdir: namespace,
      logging: {
        level: LogLevel.INFO,
      },
      isVersioned: true,
      namingStrategy: {
        ts: {
          attributeName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          className: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          enumName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          fileName: NAMING_STRATEGY_TYPE.PARAM_CASE,
          functionName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          functionParameterName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          interfaceName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          typeName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
        },
        appsync: {
          vtl: {
            file: NAMING_STRATEGY_TYPE.PARAM_CASE,
          },
        },
        attributes: {
          compositionSeperator: "#",
        },
        structures: {
          prefixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "",
            [STRUCTURE_TYPE.OUTPUT]: "",
            [STRUCTURE_TYPE.TRANSIENT]: "",
          },
          suffixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "input",
            [STRUCTURE_TYPE.OUTPUT]: "output",
            [STRUCTURE_TYPE.TRANSIENT]: "message",
          },
        },
        operations: {
          prefixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "create",
            [OPERATION_SUB_TYPE.READ_ONE]: "get",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "update",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "delete",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "import",
            [OPERATION_SUB_TYPE.LIST]: "list",
          },
          suffixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "",
            [OPERATION_SUB_TYPE.READ_ONE]: "",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "",
            [OPERATION_SUB_TYPE.LIST]: "",
          },
        },
      },
      auditStrategy: {
        create: {
          dateAttribute: {
            name: "created-at",
            shortName: "cd",
            comments: [`The date and time this record was created.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.CREATE_ONE],
          },
        },
        update: {
          dateAttribute: {
            name: "updated-at",
            shortName: "ud",
            comments: [`The date and time this record was last updated.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [
              OPERATION_SUB_TYPE.CREATE_ONE,
              OPERATION_SUB_TYPE.UPDATE_ONE,
              OPERATION_SUB_TYPE.DELETE_ONE,
            ],
          },
        },
        delete: {
          dateAttribute: {
            name: "deleted-at",
            shortName: "dd",
            comments: [`The date and time this record was deleted.`],
            type: ResourceAttributeType.DATE_TIME,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.DELETE_ONE],
            outputOn: [OPERATION_SUB_TYPE.DELETE_ONE],
          },
        },
      },
    };

    /***************************************************************************
     *
     * INIT FRACTURE PROJECT
     *
     **************************************************************************/

    // member components
    this.services = [];
    this.organizations = [];

    // project and namespace
    this.project = project;
    this.namespace = namespace;

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<FractureOptions>;

    return this;
  }

  /**
   * Build the project.
   *
   * Call this when you've configured everything, prior to preSynthesize.
   *
   * @returns void
   */
  public build() {
    this.services.forEach((service) => {
      service.build();
      new TypescriptService(service);
    });
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add a service to the fracture project.
   * @param {ServiceOptions}
   * @returns {Service}
   */
  public addService(options: ServiceOptions) {
    return new Service(this, options);
  }

  /**
   * Add an organization to the fracture project.
   * @param {OrganizationOptions}
   * @returns {Organization}
   */
  public addOrganization(options: OrganizationOptions) {
    return new Organization(this, options);
  }
}
