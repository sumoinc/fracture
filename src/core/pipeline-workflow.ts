import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { Fracture } from "./fracture";
import { Pipeline } from "./pipeline";

// const BUILD_JOBID = "build";
// const DEPLOY_JOBID_PREFIX = "deploy";
export const DEPLOY_TASK_PREFIX = "cdk:deploy";

export interface PipelineWorkflowOptions {
  pipeline: Pipeline;
}

export class PipelineWorkflow extends Component {
  // all other options
  public readonly options: PipelineWorkflowOptions;

  /**
   * Because the github workflow is build on the base project we pass in the app
   * but the component is actually based the root project
   */
  constructor(fracture: Fracture, options: PipelineWorkflowOptions) {
    super(fracture);

    this.options = options;

    // debugging output
    this.project.logger.info(
      `INIT Pipeline Workflow: "${this.pipeline.deployName}"`
    );
  }

  public get pipeline() {
    return this.options.pipeline;
  }

  preSynthesize(): void {
    const fracture = this.project as Fracture;
    const pipeline = this.pipeline;
    const app = pipeline.app;

    /*******************************************************************************
     *
     * TASK CONFIGURATION
     *
     * Adds a task to projen for each deployment stage.
     *
     ******************************************************************************/

    pipeline.stages.forEach((stage) => {
      fracture.addTask(`${DEPLOY_TASK_PREFIX}:${stage.id}`, {
        description: `CDK deploy of ${stage.name} stage`,
        exec: `cdk deploy *-${stage.id} --require-approval never`,
      });
    });

    /*******************************************************************************
     *
     * WORKFLOW CONFIGURATION
     *
     * Builds github workflow for stages created in tasks above.
     *
     ******************************************************************************/

    console.log(`DEBUG: "${pipeline.deployName}"`, fracture.name, app.name);

    const github = GitHub.of(fracture);
    if (!github) {
      throw new Error("Pipeline requires a github project");
    }

    new GithubWorkflow(github, pipeline.deployName);

    // build code

    // new BuildWorkflow(project, {
    //   name: "deploy",
    //   buildTask: project.buildTask,
    //   artifactsDirectory: project.artifactsDirectory,
    // containerImage: options.workflowContainerImage,
    // gitIdentity: this.workflowGitIdentity,
    // mutableBuild: options.mutableBuild,
    // preBuildSteps: this.renderWorkflowSetup({
    //   mutable: options.mutableBuild ?? true,
    // }),
    // postBuildSteps: options.postBuildSteps,
    // runsOn: options.workflowRunsOn,
    // workflowTriggers: options.buildWorkflowTriggers,
    // permissions: workflowPermissions,
    // });
  }
}
