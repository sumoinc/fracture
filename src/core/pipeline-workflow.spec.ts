import { synthSnapshot } from "projen/lib/util/synth";
import { Fracture } from "./fracture";
import { DEPLOY_TASK_PREFIX } from "./pipeline-workflow";
import { REGION_IDENTITIER } from "./region";

/*******************************************************************************
 * TEST SETUP
 ******************************************************************************/

let fracture: Fracture;

beforeEach(() => {
  fracture = new Fracture({
    name: "test-project",
  });
  const org = fracture.addOrganization({ id: "test-org" });
  const account = org.addAccount({ id: "id", name: "test-account" });
  const devUsEast = fracture.addEnvironment({
    account,
    region: REGION_IDENTITIER.US_EAST_1,
  });
  const devUsWest = fracture.addEnvironment({
    account,
    region: REGION_IDENTITIER.US_WEST_2,
  });
  const app = fracture.addApp({ name: "identity-service" });
  const featurePipeline = app.addPipeline({
    name: "feature-branch",
    branchTriggerPattern: "feature/*",
  });
  featurePipeline.addStage({ name: "East Coast", environment: devUsEast });
  featurePipeline.addStage({ name: "West Coast", environment: devUsWest });
});

/*******************************************************************************
 * SMOKE TEST
 ******************************************************************************/

test("Smoke test", () => {
  const workflows = synthWorkflows(fracture);
  console.log(workflows);
  expect(workflows).toBeTruthy();
});

describe("Task tests", () => {
  test("Generates expected task in projen", () => {
    const snapshot = synthSnapshot(fracture);
    const tasks = snapshot[".projen/tasks.json"].tasks;

    // filter out only deployment tasks
    const deployTasks = Object.keys(tasks)
      .filter((path) => path.startsWith(DEPLOY_TASK_PREFIX))
      .reduce((obj, key) => {
        return {
          ...obj,
          [key]: tasks[key],
        };
      }, {});

    expect(deployTasks).toMatchSnapshot();

    //console.log(JSON.stringify(deployTasks, null, 2));
  });
});

/*
describe("workflow file tests", () => {
  test("Generates expected pipeline", () => {
    const p = new TestFracture();
    const pipeline = p.testPipeline();
    p.testStage();

    const workflows = synthWorkflows(p);

    const fileContent =
      workflows[`.github/workflows/${pipeline.deployName}.yml`];
    console.log(workflows[`.github/workflows/${pipeline.deployName}.yml`]);
    expect(fileContent).toBeDefined();
    expect(fileContent).toMatchSnapshot();
  });
});

describe("formatting", () => {
  test("Deployments depend on build", () => {});
  test("Deployments depend on branch", () => {});
});
*/

/**
 *
 * Helper function to generate a workflow file so we can test it's structure.
 *
 */

function synthWorkflows(_fracture: Fracture): any {
  const snapshot = synthSnapshot(_fracture);
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(".github/workflows/"))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {});
  return filtered;
}
