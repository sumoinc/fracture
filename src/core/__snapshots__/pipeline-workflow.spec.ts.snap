// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task tests Generates expected task in projen 1`] = `
Object {
  "cdk:deploy:east-coast": Object {
    "description": "CDK deploy of East Coast stage",
    "name": "cdk:deploy:east-coast",
    "steps": Array [
      Object {
        "exec": "cdk deploy *-east-coast --require-approval never",
      },
    ],
  },
  "cdk:deploy:west-coast": Object {
    "description": "CDK deploy of West Coast stage",
    "name": "cdk:deploy:west-coast",
    "steps": Array [
      Object {
        "exec": "cdk deploy *-west-coast --require-approval never",
      },
    ],
  },
}
`;
